build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1417,156 @@ message ExecuteRequest {
   ResultsCachePolicy results_cache_policy = 8;
 }
 
+// Metadata and requests to the client for an ongoing
+// [dynamic execution][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval],
+// which will be contained in the [metadata field][google.longrunning.Operation.response] of the
+// [Operation][google.longrunning.Operation].
+//
+// Design note: Dynamic directory and file requests are made using individual
+// messages for each directory or file path. The typical implementation
+// utilizes a virtual filesystem and a directory or file read request is
+// held pending in the VFS awaiting the information from the client.
+// VFS rendering callbacks may occur in a multithreaded fashion, but this is
+// unusual, and a chatty back-and-forth request series to the client resolves
+// pending I/O faster for each thread rather than adding in a Nagle
+// algorithm implementation and using repeated fields for the dynamic requests.
+// Dynamic callbacks for content are expected to be unusual or rare by
+// construction, and inherently slow down action execution.
+message DynamicExecuteOperationMetadata {
+  // Used by the worker to request that the client upload specific content
+  // to the CAS.
+  message DynamicFileContentRequest {
+    // The file path for which the worker does not have the contents.
+    // Depending on the server and client implementation this may be a
+    // relative or fully qualified path.
+    string path = 1;
+
+    // The file hash of the content expected by the worker.
+    // If this is not the current hash of the file content at `path`
+    // then the corrected hash is returned in
+    // `DynamicExecuteOperationMetadata.dynamic_directory_metadata_response`.
+    string file_hash = 2;
+  }
+
+  // The current stage of execution.
+  ExecutionStage.Value stage = 1;
+
+  // The digest of the [Action][build.bazel.remote.execution.v2.Action]
+  // being executed.
+  Digest action_digest = 2;
+
+  // A series of incremental (streamed) stdout and stderr outputs in order of
+  // write from the action during execution. The full set of console outputs
+  // from the action will also be written in the final
+  // [ExecuteResponse][build.bazel.remote.execution.v2.ExecuteResponse].
+  // The server MAY buffer these outputs in order to return larger batches
+  // of messages less frequently. The server MAY choose to omit incremental
+  // messages and instead wait to the final `ExecuteResponse` to return all
+  // outputs from the action, however in this case the server MUST NOT return
+  // any incremental messages in this field.
+  ConsoleOutputs streamed_console_outputs = 3;
```

Comment:
My reasons for ConsoleOut inlining in the bidi protocol are different: Often we get some mega C++ compile steps (one remoted action running 500 compiler sub-processes each compiling 1 cpp file) where the incremental stdout (and occasional interleaved stderr) helps with interactive console feedback at the client from the build engine's parsing of incremental outputs. Same goes for remoting long-running things like TensorFlow training runs where incremental feedback instead of batchy final results helps with knowing the system is working.

The only alternative I could think of to get a similar user experience if mapping to streams would be to place time codes on segments of the stream so that interleaving could be restored on the client side for fidelity. That seemed nastier for no particular gain.

Any other ideas? The UX aspect of it is actually kind of important in this case, but willing to discuss alternatives. It's one of the reasons I state in the oversized doc comments here that the server might not return these, and that the final outputs (including from cached results) are replicated in the final messages. (We replay the ordered stderr+out strings to the waiting app on a cache hit too, it jjust looks like the 500 cpp files compiled very very rapidly.)